patch_subset/cbor/cbor_utils.cc

```diff
@@ -0,0 +1,360 @@
+#include "patch_subset/cbor/cbor_utils.h"
+
+namespace patch_subset::cbor {
+
+using absl::string_view;
+
+StatusCode CborUtils::GetField(const cbor_item_t& cbor_map, int field_number,
+                               cbor_item_unique_ptr& out) {
+  if (!cbor_isa_map(&cbor_map) || field_number < 0) {
+    return StatusCode::kInvalidArgument;
+  }
+  size_t size = cbor_map_size(&cbor_map);
+  cbor_pair* pairs = cbor_map_handle(&cbor_map);
+  for (size_t i = 0; i < size; i++) {
+    cbor_pair* pair = pairs + i;
+    if (!cbor_is_int(pair->key)) {
+      // Ill formed map. Keys should all be ints. Skip this field.
+      continue;
+    }
+    int32_t id;
+    StatusCode sc = DecodeInt(*pair->key, &id);
+    if (sc != StatusCode::kOk) {
+      // Ill formed map. Keys should all be valid ints. Skip this field.
+      continue;
+    }
+    if (id == field_number) {
+      // Because we are putting the pointer into a cbor_item_unique_ptr, which
+      // will eventually call cbor_decref(), the item has a new "owner".
+      cbor_item_t* value = pair->value;
+      if (value == nullptr) {
+        return StatusCode::kInvalidArgument;
+      }
+      out.reset(cbor_incref(value));
+      return StatusCode::kOk;
+    }
+  }
+  // Field not found.
+  out.reset(nullptr);
+  return StatusCode::kNotFound;
+}
+
+StatusCode CborUtils::GetUInt64Field(const cbor_item_t& map, int field_number,
+                                     optional<uint64_t>& out) {
```

Comment:
Optional:
All of these Get*Field calls follow pretty much the same pattern. I believe it would be possible to implement these internally use a templated function that would take two template parameters, the data type, and a decode function. So then this method would be implemented something like:
```
StatusCode CborUtils::GetUInt64Field(const cbor_item_t& map, int field_number,
                                     optional<uint64_t>& out) {
  return GetTypedField<uint64, DecodeUint64>(map, field_number, out);                     
}
```
That would save a considerable amount of boiler plate across these methods.

Reply:
Added a TODO